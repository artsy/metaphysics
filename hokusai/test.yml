version: "2"
services:
  metaphysics:
    command: env NODE_OPTIONS=--max_old_space_size=4096 yarn ci
    extends:
      file: build.yml
      service: metaphysics
    environment:
      - MEMCACHED_URL=metaphysics-memcached:11211
      - CI_PULL_REQUEST=$CI_PULL_REQUEST
      - CIRCLE_PULL_REQUEST=$CIRCLE_PULL_REQUEST
      - CIRCLE_BUILD_NUM=$CIRCLE_BUILD_NUM
      - ENABLE_CONSIGNMENTS_STITCHING=true
    env_file: ../.env.test
    depends_on:
      - metaphysics-memcached
  metaphysics-memcached:
    image: memcached
    ports:
      - 11211:11211
