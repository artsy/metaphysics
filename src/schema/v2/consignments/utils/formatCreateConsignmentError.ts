import {
  GraphQLInt,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
} from "graphql"
import { formatGravityError } from "lib/gravityErrorHandler"
import { ResolverContext } from "types/graphql"

export const ConsignmentInquiryMutationErrorName =
  "ConsignmentInquiryMutationError"

export const ConsignmentInquiryErrorType = new GraphQLObjectType<
  any,
  ResolverContext
>({
  name: ConsignmentInquiryMutationErrorName,
  fields: () => ({
    type: {
      type: GraphQLString,
    },
    message: {
      type: new GraphQLNonNull(GraphQLString),
    },
    error: {
      type: GraphQLString,
    },
    statusCode: {
      type: GraphQLInt,
    },
  }),
})

export const formatCreateConsignmentError = (error) => {
  return formatGravityError(error)
}
