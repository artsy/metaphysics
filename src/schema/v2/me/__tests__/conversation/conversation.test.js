/* eslint-disable promise/always-return */
import { runAuthenticatedQuery } from "schema/v2/test/utils"

describe("Me", () => {
  describe("Conversation", () => {
    const context = {
      conversationLoader: () => {
        return Promise.resolve({
          id: "420",
          initial_message: "Loved some of the works at your fair booth!",
          from_email: "collector@example.com",
          from_name: "Percy",
          _embedded: {
            last_message: {
              snippet:
                "Loved some of the works at your fair booth! About this collector: Percy is a good cat",
              from_email_address: "other-collector@example.com",
              id: "25",
              order: 1,
            },
          },
          from_last_viewed_message_id: "20",
          items: [
            {
              item_type: "Artwork",
              item_id: "artwork-42",
              title: "Pwetty Cats",
              properties: {
                title: "Pwetty Cats",
                acquireable: true,
                artists: [
                  {
                    id: "artist-42",
                  },
                ],
              },
            },
            {
              item_type: "PartnerShow",
              item_id: "show-42",
              title: "Catty Show",
              properties: {
                is_reference: true,
                display_on_partner_profile: true,
              },
            },
          ],
        })
      },
      conversationMessagesLoader: ({ sort } = { sort: "asc" }) => {
        let messages = [
          {
            id: "240",
            raw_text: "this is a good message",
            from_email_address: "fancy_german_person@posteo.de",
            from_id: null,
            attachments: [],
            metadata: {
              lewitt_invoice_id: "420i",
            },
            from: `"Percy Z" <percy@cat.com>`,
            from_principal: true,
            body: "I'm a cat",
          },
          {
            id: "241",
            raw_text: "this is a good message",
            from_email_address: "postman@posteo.de",
            from_id: null,
            attachments: [],
            metadata: {},
            from: `"Bitty Z" <Bitty@cat.com>`,
            from_principal: false,
            body: "",
          },
          {
            id: "242",
            raw_text: "this is a good message",
            from_email_address: "fancy_german_person+wunderbar@posteo.de",
            from_id: "user-42",
            attachments: [],
            metadata: {},
            from: `"Matt Z" <matt@cat.com>`,
            body: null,
          },
          {
            id: "243",
            raw_text: "this is a good message",
            from_email_address: "postman+wunderlich@posteo.de",
            from_id: "user-21",
            attachments: [],
            metadata: {},
            from: "<email@email.com>",
            from_principal: null,
            deliveries: [
              {
                opened_at: "2020-12-31T12:00:00+00:00",
              },
            ],
          },
        ]

        if (sort === "desc") {
          messages = messages.reverse()
        }

        return Promise.resolve({
          total_count: 4,
          message_details: messages,
        })
      },
    }

    it("returns a conversation", () => {
      const query = `
        {
          me {
            conversation(id: "420") {
              internalID
              initialMessage
              from {
                email
              }
              lastMessage
              unread
              messages(first: 10) {
                edges {
                  node {
                    internalID
                    isInvoice
                    invoice {
                      id
                      internalID
                      paymentURL
                      state
                      total
                    }
                    isFromUser
                    from {
                      name
                      email
                    }
                    body
                    deliveries {
                      openedAt
                    }
                  }
                }
              }
            }
          }
        }
      `

      return runAuthenticatedQuery(query, context).then(
        ({ me: conversation }) => {
          expect(conversation).toMatchSnapshot()
        }
      )
    })

    describe("concerning unread indicator", () => {
      it("returns the right unread status", () => {
        const query = `
          {
            me {
              conversation(id: "420") {
                isLastMessageToUser
                unread
                lastMessageDeliveryID
              }
            }
          }
        `

        const message = {
          message_details: [
            {
              deliveries: [
                {
                  id: "2",
                  email: "collector@example.com",
                  opened_at: "2020-12-31T12:00:00+00:00",
                },
              ],
            },
          ],
        }

        const customRootValue = Object.assign({}, context, {
          conversationMessagesLoader: () => Promise.resolve(message),
        })

        return runAuthenticatedQuery(query, customRootValue).then(
          ({ me: { conversation } }) => {
            expect(conversation).toMatchSnapshot()
          }
        )
      })
    })

    describe("concerning items", () => {
      const query = `
        {
          me {
            conversation(id: "420") {
              items {
                title
                item {
                  __typename
                  ... on Artwork {
                    isAcquireable
                  }
                  ... on Show {
                    isReference
                  }
                }
              }
            }
          }
        }
      `

      it("returns the conversation items", () => {
        return runAuthenticatedQuery(query, context).then(
          ({
            me: {
              conversation: { items },
            },
          }) => {
            expect(items.length).toEqual(2)
            expect(items).toMatchSnapshot()
          }
        )
      })

      it("doesnt return invalid items", () => {
        const customRootValue = Object.assign({}, context, {
          // Get a copy of the conversation payload, mutate it, and return that instead.
          conversationLoader: () =>
            context.conversationLoader().then((conversation) => {
              const convo = conversation
              convo.items[0].properties = {}
              return convo
            }),
        })

        return runAuthenticatedQuery(query, customRootValue).then(
          ({
            me: {
              conversation: { items },
            },
          }) => {
            expect(items.length).toEqual(1)
            expect(items).toMatchSnapshot()
          }
        )
      })
    })

    describe("messages", () => {
      const getQuery = (sort = "ASC") => {
        return `
          {
            me {
              conversation(id: "420") {
                messages(first: 10, sort: ${sort}) {
                  edges {
                    node {
                      internalID
                      isFromUser
                    }
                  }
                }
              }
            }
          }
        `
      }

      it("returns messages in ascending order", () => {
        const query = getQuery()

        return runAuthenticatedQuery(query, context).then(
          ({
            me: {
              conversation: { messages },
            },
          }) => {
            expect(messages.edges.length).toEqual(4)
            expect(messages.edges[0].node.internalID).toEqual("240")
          }
        )
      })

      it("returns messages in descending order", () => {
        const query = getQuery("DESC")

        return runAuthenticatedQuery(query, context).then(
          ({
            me: {
              conversation: { messages },
            },
          }) => {
            expect(messages.edges.length).toEqual(4)
            expect(messages.edges[0].node.internalID).toEqual("243")
          }
        )
      })
      it("returns proper is_from_user for different values of `from_principal`", () => {
        const query = getQuery()

        return runAuthenticatedQuery(query, context).then(
          ({
            me: {
              conversation: { messages },
            },
          }) => {
            expect(messages.edges.length).toEqual(4)
            expect(messages.edges[0].node.isFromUser).toEqual(true)
            expect(messages.edges[1].node.isFromUser).toEqual(false)
            expect(messages.edges[2].node.isFromUser).toEqual(true)
            expect(messages.edges[3].node.isFromUser).toEqual(false)
          }
        )
      })
    })
  })
})
