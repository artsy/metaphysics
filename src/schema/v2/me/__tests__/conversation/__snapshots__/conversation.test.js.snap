// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Me Conversation concerning items doesnt return invalid items 1`] = `
Array [
  Object {
    "item": Object {
      "__typename": "Show",
      "isReference": true,
    },
    "title": "Catty Show",
  },
]
`;

exports[`Me Conversation concerning items returns the conversation items 1`] = `
Array [
  Object {
    "item": Object {
      "__typename": "Artwork",
      "isAcquireable": true,
    },
    "title": "Pwetty Cats",
  },
  Object {
    "item": Object {
      "__typename": "Show",
      "isReference": true,
    },
    "title": "Catty Show",
  },
]
`;

exports[`Me Conversation concerning unread indicator returns the right unread status 1`] = `
Object {
  "isLastMessageToUser": true,
  "lastMessageDeliveryID": "2",
  "unread": true,
}
`;

exports[`Me Conversation returns a conversation 1`] = `
Object {
  "conversation": Object {
    "from": Object {
      "email": "collector@example.com",
    },
    "initialMessage": "Loved some of the works at your fair booth!",
    "internalID": "420",
    "lastMessage": "Loved some of the works at your fair booth!",
    "messages": Object {
      "edges": Array [
        Object {
          "node": Object {
            "body": "I'm a cat",
            "deliveries": null,
            "from": Object {
              "email": "fancy_german_person@posteo.de",
              "name": "Percy Z",
            },
            "internalID": "240",
            "invoice": null,
            "isFromUser": true,
            "isInvoice": true,
          },
        },
        Object {
          "node": Object {
            "body": "",
            "deliveries": null,
            "from": Object {
              "email": "postman@posteo.de",
              "name": "Bitty Z",
            },
            "internalID": "241",
            "invoice": null,
            "isFromUser": false,
            "isInvoice": false,
          },
        },
        Object {
          "node": Object {
            "body": null,
            "deliveries": null,
            "from": Object {
              "email": "fancy_german_person+wunderbar@posteo.de",
              "name": "Matt Z",
            },
            "internalID": "242",
            "invoice": null,
            "isFromUser": true,
            "isInvoice": false,
          },
        },
        Object {
          "node": Object {
            "body": null,
            "deliveries": Array [
              Object {
                "openedAt": "2020-12-31T12:00:00+00:00",
              },
            ],
            "from": Object {
              "email": "postman+wunderlich@posteo.de",
              "name": null,
            },
            "internalID": "243",
            "invoice": null,
            "isFromUser": false,
            "isInvoice": false,
          },
        },
      ],
    },
    "unread": true,
  },
}
`;
